File: 3.longest-substring-without-repeating-characters.java
// @leet start
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> charCount = new HashMap<>();
        char strChar[] = s.toCharArray();
        int result = 0;
        int l = 0;

        for (int r = 0; r < s.length(); r++) {
            char curr = strChar[r];
            charCount.put(curr, charCount.getOrDefault(curr, 0) + 1);
            while (charCount.get(curr) > 1) {
                char left = strChar[l];
                charCount.put(left, charCount.getOrDefault(left, 0) - 1);
                l++;
            }
            result = Math.max(result, r - l + 1);
        }
        
        return result;
    }
}
// @leet end

File: 3.longest-substring-without-repeating-characters.ts
// @leet start
function lengthOfLongestSubstring(s: string): number {
	if (s.length <= 1) {
		return s.length;
	}
	let result = 0;
	let dict = {};
	let back_ptr = 0;

	dict[s[0]] = 1;

	for (let front_ptr = 1; front_ptr < s.length; front_ptr++) {
		while (dict[s[front_ptr]] == 1) {
			dict[s[back_ptr++]] = 0;
		}

		dict[s[front_ptr]] = 1;

		result = Math.max(result, front_ptr - back_ptr + 1);
	}

	return result;
    
};
// @leet end
